File: src/main/java/org/academy/api/client/render/MsdfUniformData.java
package org.academy.api.client.render;

import com.mojang.blaze3d.buffers.Std140Builder;
import com.mojang.blaze3d.buffers.Std140SizeCalculator;
import net.minecraft.client.renderer.DynamicUniformStorage;
import org.joml.Vector4f;

import java.nio.ByteBuffer;

public record MsdfUniformData(
        float range,
        float thickness,
        float outlineThickness,
        Vector4f outlineColor
) implements DynamicUniformStorage.DynamicUniform {
    public static final int UBO_SIZE = new Std140SizeCalculator()
            .putFloat()
            .putFloat()
            .putFloat()
            .putVec4()
            .get();

    @Override
    public void write(ByteBuffer buffer) {
        Std140Builder.intoBuffer(buffer)
                .putFloat(range)
                .putFloat(thickness)
                .putFloat(outlineThickness)
                .putVec4(outlineColor);
    }
}

File: src/main/kotlin/org/academy/api/client/render/MsdfUniformData.kt
package org.academy.api.client.render

import com.mojang.blaze3d.buffers.Std140Builder
import com.mojang.blaze3d.buffers.Std140SizeCalculator
import net.minecraft.client.renderer.DynamicUniformStorage.DynamicUniform
import org.joml.Vector4f
import java.nio.ByteBuffer

data class MsdfUniformData(
    val range: Float,
    val thickness: Float,
    val outlineThickness: Float,
    val outlineColor: Vector4f
) : DynamicUniform {
    override fun write(buffer: ByteBuffer) {
        Std140Builder.intoBuffer(buffer)
            .putFloat(range)
            .putFloat(thickness)
            .putFloat(outlineThickness)
            .putVec4(outlineColor)
    }

    companion object {
        val UBO_SIZE: Int = Std140SizeCalculator()
            .putFloat()
            .putFloat()
            .putFloat()
            .putVec4()
            .get()
    }
}
